In [2]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [3]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
lendingDF = pd.read_csv(Path("Resources", "lending_data.csv"))

# Review the DataFrame
lendingDF.info()
lendingDF.head(1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77536 entries, 0 to 77535
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   loan_size         77536 non-null  float64
 1   interest_rate     77536 non-null  float64
 2   borrower_income   77536 non-null  int64  
 3   debt_to_income    77536 non-null  float64
 4   num_of_accounts   77536 non-null  int64  
 5   derogatory_marks  77536 non-null  int64  
 6   total_debt        77536 non-null  int64  
 7   loan_status       77536 non-null  int64  
dtypes: float64(3), int64(5)
memory usage: 4.7 MB


loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [29]:
# Separate the data into labels and features
# Separate the y variable, the labels
labels = lendingDF["loan_status"]
y = labels

# Separate the X variable, the features
features = lendingDF[["loan_size", "interest_rate", "borrower_income", "debt_to_income", "num_of_accounts", "derogatory_marks", "total_debt"]]
x = features

In [30]:
# Review the y variable Series
y.describe()
print(f"row length {len(labels)}")

row length 77536


In [31]:
# Review the X variable DataFrame
x = pd.DataFrame(features)
print(f"row length {len(features)}")
x.describe()

row length 77536


loan_size  interest_rate  borrower_income  debt_to_income  \
count  77536.000000   77536.000000     77536.000000    77536.000000   
mean    9805.562577       7.292333     49221.949804        0.377318   
std     2093.223153       0.889495      8371.635077        0.081519   
min     5000.000000       5.250000     30000.000000        0.000000   
25%     8700.000000       6.825000     44800.000000        0.330357   
50%     9500.000000       7.172000     48100.000000        0.376299   
75%    10400.000000       7.528000     51400.000000        0.416342   
max    23800.000000      13.235000    105200.000000        0.714829   

       num_of_accounts  derogatory_marks    total_debt  
count     77536.000000      77536.000000  77536.000000  
mean          3.826610          0.392308  19221.949804  
std           1.904426          0.582086   8371.635077  
min           0.000000          0.000000      0.000000  
25%           3.000000          0.000000  14800.000000  
50%           4.000000          0.000000  18100.000000  
75%           4.000000          1.000000  21400.000000  
max          16.000000          3.000000  75200.000000

### Step 3: Split the data into training and testing datasets by using `train_test_split`.

In [32]:
# Import the train_test_learn module  #stratify?
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
x_train, x_test, y_train, y_test = train_test_split(features, labels, random_state = 1)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [33]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
classifier = LogisticRegression(solver='lbfgs', random_state=1)
classifier

# Fit the model using training data
classifier.fit(x_train, y_train)

print(f"Training Data Score {classifier.score(x_train, y_train)}")
print(f"Testing Data Score {classifier.score(x_test, y_test)}")

Training Data Score 0.9929323153115972
Testing Data Score 0.9924680148576145


### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [38]:
# Make a prediction using the testing data
predictions = classifier.predict(x_test)
predictions_DF = pd.DataFrame({"Predictions": predictions, "Actual": y_test})
predictions_DF.value_counts()

Predictions  Actual
0            0         18655
1            1           583
             0           110
0            1            36
Name: count, dtype: int64

### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [39]:
# Generate a confusion matrix for the model
confusion_matrix(y_test, predictions)

array([[18655,   110],
       [   36,   583]], dtype=int64)

In [41]:
# Print the classification report for the model
outcomes = ["healthy", "high_risk"]
print(classification_report(y_test, predictions, target_names = outcomes))

              precision    recall  f1-score   support

     healthy       1.00      0.99      1.00     18765
   high_risk       0.84      0.94      0.89       619

    accuracy                           0.99     19384
   macro avg       0.92      0.97      0.94     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** Assuming my code is right, the predictions were spot-on for healthy loans. The worst score the model produced was the 84% precision predicting high risk loans. I'm not sure if that means it predicted more false positives or fewer positives than the actual number, but I think this stems from having fewer actual high risk loans in the dataset. Essentially, the model was able to train & identify the patterns for healthy loans more than high risk ones.

---